src/BenchmarksApps/SslStream/Client/Program.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Diagnostics;
+using System.Net;
+using System.Net.Security;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Net.Sockets;
+using System.CommandLine;
+using Microsoft.Crank.EventSources;
+using SslStreamCommon;
+using SslStreamClient;
+
+// avoid false sharing among counters
+[StructLayout(LayoutKind.Explicit)]
+internal struct Counters
+{
+    [FieldOffset(0)] public long BytesRead;
+    [FieldOffset(64)] public long BytesWritten;
+    [FieldOffset(128)] public long TotalHandshakes;
+
+    public void Reset()
+    {
+        BytesRead = 0;
+        BytesWritten = 0;
+        TotalHandshakes = 0;
+    }
+}
+
+internal class Program
+{
+    private static Counters s_counters;
+    private static bool s_isRunning;
+    private static bool s_isWarmup;
+
+    private static async Task<int> Main(string[] args)
+    {
+        var rootCommand = new RootCommand("SslStream benchmark client");
+        OptionsBinder.AddOptions(rootCommand);
+        rootCommand.SetHandler<ClientOptions>(Run, new OptionsBinder());
+        return await rootCommand.InvokeAsync(args).ConfigureAwait(false);
+    }
+
+    static async Task<int> Run(ClientOptions options)
+    {
+        SetupMeasurements();
+
+        try
+        {
+            switch (options.Scenario)
+            {
+                case Scenario.ReadWrite:
+                    await RunReadWriteScenario(options).ConfigureAwait(false);
+                    break;
+                case Scenario.Handshake:
+                    await RunHandshakeScenario(options).ConfigureAwait(false);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Unknown scenario: {options.Scenario}");
+            }
+        }
+        catch (Exception e)
+        {
+            Log($"Exception occured: {e}");
+            return 1;
+        }
+
+        return 0;
+    }
+
+    static async Task RunHandshakeScenario(ClientOptions options)
+    {
+        BenchmarksEventSource.Register("sslstream/handshake/mean", Operations.Avg, Operations.Avg, "Mean handshake duration (ms)", "Handshakes duration in milliseconds - mean", "n3");
+
+        static async Task DoRun(ClientOptions options)
+        {
+            while (s_isRunning)
+            {
+                using var sock = new Socket(SocketType.Stream, ProtocolType.Tcp);
+                await sock.ConnectAsync(options.Hostname, options.Port).ConfigureAwait(false);
+                await using var stream = await EstablishSslStreamAsync(sock, options).ConfigureAwait(false);
+                s_counters.TotalHandshakes++;
+            }
+        }
+
+        s_isRunning = true;
+        var task = Task.Run(() => DoRun(options));
+        await Task.Delay(options.Warmup).ConfigureAwait(false);
+        s_isRunning = false;
```

Comment:
Hmmm this wasn't _exactly_ what I had in mind, I remember that the idea behind the HttpClient benchmark approach was to not actually _stop_ between warmup and the "real" metric collection -- so that any additional cost of creating/growing a state machine etc will not affect the metrics (that being said, I don't know _exactly_ how much it affects the metrics; we had this more as a theoretical discussion with Geoff and Cory when I did the initial work on HttpClient benchmarks)

We can leave it as it is right now, I don't want to hold this back much longer, but we might want to address this later.

Reply:
(now I see that read-write scenarios seem to follow the pattern from HttpClient benchmarks, but handshakes don't 😅)

Reply:
Converted this to use similar approach as in httpclient benchmarks, and even made it report percentiles